### PR TITLE
if users are None/empty, dont assume sameness

### PR DESCRIPTION
**Pull request by bcoca**
prevents None == None cases in which 'defaults' for remote and become differ

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

actions

### CODE REVIEW

**Comment by sivel** [comment]
We should probably extend the tests in `test_action_base_sudo_only_if_user_differs` at `test/units/plugins/action/test_action.py` to test this logic.

**Comment by bcoca** [comment]
and add clog, was just seeing if all current tests passed

File: lib/ansible/plugins/action/__init__.py [hunk 1]
```diff
@@ -1039,9 +1039,11 @@ def _low_level_execute_command(self, cmd, sudoable=True, in_data=None, executabl
             display.debug("_low_level_execute_command(): changing cwd to %s for this command" % chdir)
             cmd = self._connection._shell.append_command('cd %s' % chdir, cmd)
 
-        if (sudoable and self._connection.transport != 'network_cli' and self._connection.become and
-                (C.BECOME_ALLOW_SAME_USER or
-                 self.get_become_option('become_user') != self._get_remote_user())):
+        ruser = self._get_remote_user()
+        buser = self.get_become_option('become_user')
+        if (sudoable and self._connection.become and  # if sudoable and have become
```
Comment (samdoran) [inline lib/ansible/plugins/action/__init__.py:1044]: This conditional is complex enough that I think it would be helpful to have the logic broken out and in its own method.
```suggestion
        if self._should_i_become(sudoable, buser, ruser):
```

```python
def _should_i_become(self, sudoable, buser, ruser):
    # Should we become?
    if not sudoable:
        return False

    # Does the connection support become?
    if not self._connection.become or self._connection.transport == 'network_cli':
        return False

    # Corner case of trying to become the same user
    if ruser or buser and (buser != ruser):
        # Only allow becoming same user if config says so
        if not C.BECOME_ALLOW_SAME_USER:
            return False

    return True
```

**Review by samdoran — CHANGES_REQUESTED** [review]
We definitely need test cases for this. It's pretty hard to follow this logic even when it is broken out into a separate method.

File: lib/ansible/plugins/action/__init__.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/plugins/action/__init__.py:1044]: at 3 conditions that are not reused, i still think it doesn't merit it's own function.

Comment (samdoran) [reply lib/ansible/plugins/action/__init__.py:1044]: It could still be broken out without being a separate function. Having a separate function reads easier, but I am mostly suggesting the logic be broken out more for easier reading and maintenance.

Comment (bcoca) [reply lib/ansible/plugins/action/__init__.py:1044]: i would prefer not to refactor at the same time as i change the condition ... tends to lead to more changes in behaviour than desired.

**Review by samdoran — APPROVED** [review]
